### PR TITLE
Licence/copyright notice at the top of source files

### PR DESCRIPTION
**Pull request by acabarbaye**
### Description
Creates a utility for putting a licence notice at the top of every file.
This is complementary to #15 in order to fulfill OpenChain certification.

Note:
Left to do,
- run on all projects
- add to .pre-commit.yml in cookie cutter template
- Ensure every project specifies `PROJECT_NAME` in pyproject.toml

### Test Coverage
- [ ]  This change is covered by existing or additional automated tests.
- [x]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [ ]  Additional tests are not required for this change (e.g. documentation update).

### CODE REVIEW

File: mbed_tools_ci_scripts/license_files.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            author="${owner}",
+            date=" ${years}",
+        )
+        with tempfile.NamedTemporaryFile(suffix=".tmpl", delete=False) as template_file:
+            template_file_path = Path(template_file.name)
+            logger.debug(f"Creates template file in {str(template_file_path)}")
+            template_file.write(template_string.encode("utf8"))
+            template_file.close()
+            copyright_config = self.get_tool_config(template_file_path)
+            self.run_licensehearders(copyright_config, verbose_count)
```
Comment (rwalton-arm) [inline mbed_tools_ci_scripts/license_files.py]: Should this be `run_licenseheaders`?

File: mbed_tools_ci_scripts/license_files.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    def run_licensehearders(self, config: dict, verbose_count: int) -> None:
```
Comment (rwalton-arm) [inline mbed_tools_ci_scripts/license_files.py]: Should this be `run_licenseheaders`?

File: mbed_tools_ci_scripts/license_files.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        """Runs licenseheaders tool."""
+        args = ["licenseheaders"]
+        args.extend([item for pair in {f"--{k}": str(v) for (k, v) in config.items()}.items() for item in pair])
```
Comment (rwalton-arm) [inline mbed_tools_ci_scripts/license_files.py]: Any way to simplify this expression? It's a bit complicated. Maybe just a regular `for` loop would be clearer?

File: mbed_tools_ci_scripts/license_files.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        subprocess.check_call(args)
+
+    def get_tool_config(self, template_file: Path) -> dict:
+        """Gets the configuration for licenseheaders."""
+        return {
+            "owner": configuration.get_value(ConfigurationVariable.ORGANISATION),
+            "dir": configuration.get_value(ConfigurationVariable.PROJECT_ROOT),
+            "projname": configuration.get_value(ConfigurationVariable.PROJECT_NAME),
+            "tmpl": str(template_file),
+            "years": f"2020-{datetime.now().year}",
```
Comment (rwalton-arm) [inline mbed_tools_ci_scripts/license_files.py]: Does it make sense for this to come out as "2020-2020"? Should it just be "2020" if `datetime.now.year` is the same year?

File: mbed_tools_ci_scripts/license_files.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+SPDX-License-Identifier: {licence_identifier}
+"""
+
+
+class CopyrightNotifierToolWrapper:
+    """Wrapper over the licenseheaders tool.
+
+    See https://github.com/johann-petrak/licenseheader for more details about the tool in use.
+    copyright (https://github.com/knipknap/copyright) was first considered but
+    comprises quite a few bugs and does not seem active anymore.
```
Comment (rosiemiller2) [inline mbed_tools_ci_scripts/license_files.py]: Is this the right place to put this? Feels more like an architectural decision to me, we need a record, but users don't necessarily need to know.

File: mbed_tools_ci_scripts/license_files.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+from pathlib import Path
+
+logger = logging.getLogger(__name__)
+
+LICENCE_HEADER_TEMPLATE = """Copyright (c){date}, {author}. All rights reserved.
+SPDX-License-Identifier: {licence_identifier}
+"""
+
+
+class CopyrightNotifierToolWrapper:
```
Comment (rwalton-arm) [inline mbed_tools_ci_scripts/license_files.py]: I'm not sure about the naming here. Do we need to include the word "wrapper"? What's wrong with just calling this `LicenseHeader`?

File: mbed_tools_ci_scripts/license_files.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    """Wrapper over the licenseheaders tool.
+
+    See https://github.com/johann-petrak/licenseheader for more details about the tool in use.
```
Comment (rosiemiller2) [inline mbed_tools_ci_scripts/license_files.py]: The link leads to a 404, I think it should be https://github.com/johann-petrak/licenseheaders (its just missing an s off the end).

File: mbed_tools_ci_scripts/license_files.py (hunk 6 above)
Comment (acabarbaye) [reply mbed_tools_ci_scripts/license_files.py]: I changed this

File: mbed_tools_ci_scripts/license_files.py (hunk 7 above)
Comment (acabarbaye) [reply mbed_tools_ci_scripts/license_files.py]: thanks for checking. Was the result of a bad copy/paste

File: mbed_tools_ci_scripts/license_files.py (hunk 5 above)
Comment (acabarbaye) [reply mbed_tools_ci_scripts/license_files.py]: moved this out of the docstring

File: mbed_tools_ci_scripts/license_files.py (hunk 4 above)
Comment (acabarbaye) [reply mbed_tools_ci_scripts/license_files.py]: changed this

File: mbed_tools_ci_scripts/license_files.py (hunk 3 above)
Comment (acabarbaye) [reply mbed_tools_ci_scripts/license_files.py]: I broke it down

File: mbed_tools_ci_scripts/license_files.py [hunk 8]
```diff
@@ -15,54 +15,71 @@
...
         """Runs licenseheaders tool."""
         args = ["licenseheaders"]
-        args.extend([item for pair in {f"--{k}": str(v) for (k, v) in config.items()}.items() for item in pair])
+        args_dict = {f"--{k}": str(v) for (k, v) in config.items()}
+        args.extend(flatten_dictionary(args_dict))
         if verbose_count > 0:
             args.append(f"-{''.join(['v'] * verbose_count)}")
         subprocess.check_call(args)
 
+    def _determines_copyright_dates(self) -> str:
```
Comment (rwalton-arm) [inline mbed_tools_ci_scripts/license_files.py]: Should this be a `staticmethod` or better yet, a free function? `self` isn't used here.

File: mbed_tools_ci_scripts/license_files.py [hunk 9]
```diff
@@ -15,54 +15,71 @@
...
-    def run_licensehearders(self, config: dict, verbose_count: int) -> None:
+    def _generate_header_template(self) -> str:
+        """Generates the header template which is put at the top of source files."""
+        return LICENCE_HEADER_TEMPLATE.format(
+            licence_identifier=configuration.get_value(ConfigurationVariable.FILE_LICENCE_IDENTIFIER),
+            author="${owner}",
+            date="${years}",
+        )
+
+    def _call_licensehearders(self, config: dict, verbose_count: int) -> None:
```
Comment (rwalton-arm) [inline mbed_tools_ci_scripts/license_files.py]: Should this be a `staticmethod` or better yet, a free function? `self` isn't used here.

File: mbed_tools_ci_scripts/license_files.py [hunk 10]
```diff
@@ -15,54 +15,71 @@
...
 
-LICENCE_HEADER_TEMPLATE = """Copyright (c){date}, {author}. All rights reserved.
+LICENCE_HEADER_TEMPLATE = """Copyright (c) {date}, {author}. All rights reserved.
 SPDX-License-Identifier: {licence_identifier}
 """
 
 
-class CopyrightNotifierToolWrapper:
-    """Wrapper over the licenseheaders tool.
+class LicenceHeader:
```
Comment (rwalton-arm) [inline mbed_tools_ci_scripts/license_files.py]: Sorry I didn't mention this earlier but now we've added more static methods I wonder if this should be a class at all? Seems like we have a function that just calls some other functions, there's no invariant or state we're managing here.

Comment (acabarbaye) [reply mbed_tools_ci_scripts/license_files.py]: true
changed that all

**Pull request merged**